### PR TITLE
Issue #9315: JavadocTagContinuationIndentation does not check text inside HTML elements

### LINKED ISSUE #9315: JavadocTagContinuationIndentation does not check text inside HTML elements
From https://github.com/checkstyle/checkstyle/pull/8177#discussion_r579386002

Check documentation: https://checkstyle.org/config_javadoc.html#JavadocTagContinuationIndentation

```cmd
D:\checkstyletest>type config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="severity" value="error"/>
    <module name="TreeWalker">
                  <module name="JavadocTagContinuationIndentation"/>
    </module>
</module>

D:\checkstyletest>type test\MyClass.java
class MyClass {

    /**
     * Javadoc.
     *
     * @param x This is the start of a param tag.
     *     Continuing lines must have indentation. // ok
     * Lines without indentation are violations. // violation
     * <p>HTML tags are also affected by this rule.</p> // not a violation yet, will be fixed in #8177
     * And subsequent text is also checked. // violation
     *     <p>
     * However, text nested inside HTML elements are unaffected. // no violation
     *     </p>
     */
    void foo(int x) {
        // do stuff
    }

}
D:\checkstyletest>set RUN_LOCALE="-Duser.language=en -Duser.country=US"

D:\checkstyletest>java -jar "-Duser.language=en -Duser.country=US" checkstyle-8.40-all.jar -c config.xml test\MyClass.java
Starting audit...
[ERROR] D:\checkstyletest\test\MyClass.java:8: Line continuation have incorrect indentation level, expected level should be 4. [JavadocTagContinuationIndentation]
[ERROR] D:\checkstyletest\test\MyClass.java:10: Line continuation have incorrect indentation level, expected level should be 4. [JavadocTagContinuationIndentation]
Audit done.
Checkstyle ends with 2 errors.
```

---

The line `However, text nested inside HTML elements are unaffected.` does not raise any violation, despite having no indentation with respect to the `@param` tag. In general, when HTML elements are involved the text between the start and end tags is not checked.

This behavior may be unintuitive as the check documentation does not have any examples on whether or not this is considered a violation.
- ~~If this is intended behavior, then the documentation should be updated to include a similar example so that future users will not be confused by this case.~~ see https://github.com/checkstyle/checkstyle/issues/9315#issuecomment-808765706
- Otherwise, if this is considered a bug then please see https://github.com/checkstyle/checkstyle/pull/8177#discussion_r579386002 for details on how it may be fixed.

### PR DESCRIPTION
**Pull request by shashwatj07**
Resolves #9315 

Existing UT already cover added cases

Diff Regression config: https://gist.githubusercontent.com/shashwatj07/4b4387bc73a204c6d59dda2f7dc9fd2a/raw/6fa2a34b3c524ac8cc25f3b454449c4bc84dbed2/config.xml
Diff Regression projects: https://gist.githubusercontent.com/shashwatj07/b3de11a5c370e5f6092e31776b9c3b91/raw/8f42d25090f9d1b10c9b5d56ba39b482ac0aa91f/projects.properties

### CODE REVIEW

**Comment by romani** [comment]
Hm, there is  problem in our satellite project
```
[INFO] --- maven-checkstyle-plugin:3.1.2:check (default) @ sevntu-checks ---
[INFO] There is 1 error reported by Checkstyle 8.42-SNAPSHOT with ../../../config/checkstyle_checks.xml ruleset.
[ERROR] src/main/java/com/github/sevntu/checkstyle/checks/coding/AvoidHidingCauseExceptionCheck.java:[168] (javadoc) JavadocTagContinuationIndentation: Line continuation have incorrect indentation level, expected level should be 4.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
```

**Comment by romani** [comment]
As it is functionional update of main code we need regression diff report, please generate.

**Comment by shashwatj07** [comment]
> Please send PR to fix issue at https://github.com/sevntu-checkstyle/sevntu.checkstyle/blob/master/sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/AvoidHidingCauseExceptionCheck.java
> It should be merged first

https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/835

**Comment by romani** [comment]
PR is merged, wercker is restarted

**Comment by shashwatj07** [comment]
> You can reproduce problem by:
> Command: `./.ci/wercker.sh no-error-sevntu-checks`

https://gist.github.com/shashwatj07/bd66b9a61602aff947b7bad5c3393795
Not sure why this error. Tried several times.

**Comment by shashwatj07** [comment]
GitHub, generate report

**Comment by shashwatj07** [comment]
GitHub, generate report

**Comment by shashwatj07** [comment]
@romani ping

**Comment by shashwatj07** [comment]
@timurt please review

**Review by timurt — APPROVED** [review]

**Review by pbludov — DISMISSED** [review]

**Comment by pbludov** [comment]
@romani this one looks like a false positive: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/876ad3a_2021071907/reports/diff/hibernate-orm/index.html#A1

This is just a text inside `<i>....</i>` it is too much to require indentation in this case.

**Comment by shashwatj07** [comment]
@romani please have a look

**Comment by romani** [comment]
>  this one looks like a false positive:

it is forse-positive, we need to avoid this violation as user have no way to fix it.
@shashwatj07 , please fix it.

**Comment by shashwatj07** [comment]
> @shashwatj07 , please fix it.

Should we maintain a lot of such tags where we have to ignore violation?
Maybe all of these? https://www.w3schools.com/html/html_formatting.asp
@romani @pbludov

**Comment by pbludov** [comment]
@shashwatj07 note that for this case
```
 * to the {@link EntityManagerFactory} as outlined in the JPA specification, section <i>7.3 Obtaining an Entity
135  * Manager Factory</i> based on runtime environment.  Additionally this class will be removed in Hibernate release
```
I think there should be no violation, since the tag `<i>` is deep inside the text. For normal cases like
```
 * <ul> // violation, not correctly indented
 * <li> // ok, inside a HTML element
 * cvs package // ok, inside a HTML element
 * </li> // ok, inside a HTML element
 * <li> // ok, inside a HTML element
```
it is ok to require some indentation. So the rule should be like that:

If there is nothing but whitespace in front of a HTML tag, then the indentation should be forced.

@timurt @romani please share your thoughts.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTagContinuationIndentationCheck.java [hunk 1]
```diff
@@ -225,6 +225,10 @@ else if (!CommonUtil.isBlank(text.substring(1, offset + 1))) {
         final List<DetailNode> textNodes = new ArrayList<>();
         DetailNode node = JavadocUtil.getFirstChild(descriptionNode);
         while (JavadocUtil.getNextSibling(node) != null) {
+            if (node.getType() == JavadocTokenTypes.HTML_ELEMENT) {
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTagContinuationIndentationCheck.java:228]: Cases like "some text <i>word</i> more text" should not trigger he violation.

**Review by pbludov — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTagContinuationIndentationCheck.java (hunk 1 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTagContinuationIndentationCheck.java:228]: This Check should have nothing to do with HTML tags
https://checkstyle.org/config_javadoc.html#JavadocTagContinuationIndentation
all content of javadoc block tags is simple text that should be intended +4 to visually show javadoc tags

```
/**
 * @tag comment
 *     Indentation spacing is 4.  <p> OK
 *     Indentation spacing is 4. OK </p>
 *     Indentation spacing is 4. OK
 * @param comment
 *     Indentation spacing is 4.  <p> OK
 *     Indentation spacing is 4. OK </p>
 *     Indentation spacing is 4. OK
 */
public class Test {
}
```

Comment (shashwatj07) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTagContinuationIndentationCheck.java:228]: We are not flagging violation on HTML tags. We are just traversing deep inside tags to get ```LEADING_ASTERISK``` tokens which we call newline nodes. Then we check if that line is indented or not.

> Cases like "some text word more text" should not trigger he violation.

they won't trigger violation if the line is indented correctly. I have added it at the end of our test file.

> @romani this one looks like a false positive: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/876ad3a_2021071907/reports/diff/hibernate-orm/index.html#A1
> 
> This is just a text inside `<i>....</i>` it is too much to require indentation in this case.

this is not a false positive as the line isn't indented correctly and violation is justified. Earlier the ```LEADING_ASTERISK```  inside the HTML tag wasn't flagged leading to a false negative.

**Comment by romani** [comment]
GitHub, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTagContinuationIndentationCheck.java (hunk 1 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTagContinuationIndentationCheck.java:228]: in code change I see now:
```
     * some text <i>word</i> more text. // violation
     *     some text <i>word</i> more text. // ok
```

so case from @pbludov  is covered.

**Comment by romani** [comment]
@shashwatj07 ,

https://github.com/spring-projects/spring-integration use "main" branch instead of master now, please update your projects.properties file and generate report.
Please recheck report results for false positives.

**Comment by shashwatj07** [comment]
GitHub, generate report

**Review by romani — APPROVED** [review]
Ok to merge

**Comment by romani** [comment]
@shashwatj07 , please resolve conflict

**Comment by shashwatj07** [comment]
> @shashwatj07 , please resolve conflict

done

**Review by pbludov — APPROVED** [review]

**Pull request merged**